apply/local/applier.go

```diff
@@ -0,0 +1,333 @@
+			if err != nil {
+				return err
+			}
+			if clusterBeforeUpgrade.Spec.Image != c.ClusterDesired.Spec.Image {
+				c.ClusterCurrent.Spec.Image = clusterBeforeUpgrade.Spec.Image
+			}
+		}
+	}
+	// all commands need to do
+	err = utils.SaveClusterfile(c.ClusterDesired)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Applier) upgradeCluster() error {
+	var todoList []ActionName
+	todoList = append(todoList, PullIfNotExist)
+	todoList = append(todoList, MountImage)
+	todoList = append(todoList, MountRootfs)
+	todoList = append(todoList, Upgrade)
+	if len(c.MastersToJoin) > 0 || len(c.MastersToDelete) > 0 {
+		todoList = append(todoList, ApplyMasters)
+	}
+	if len(c.NodesToJoin) > 0 || len(c.NodesToDelete) > 0 {
+		todoList = append(todoList, ApplyNodes)
+	}
+	todoList = append(todoList, Guest)
+	todoList = append(todoList, UnMountImage)
+	return c.run(todoList)
+}
+
+func (c *Applier) scaleCluster() error {
+	var todoList []ActionName
+	todoList = append(todoList, PullIfNotExist)
+	todoList = append(todoList, MountImage)
+	todoList = append(todoList, MountRootfs)
+	if len(c.MastersToJoin) > 0 || len(c.MastersToDelete) > 0 {
+		todoList = append(todoList, ApplyMasters)
+	}
+	if len(c.NodesToJoin) > 0 || len(c.NodesToDelete) > 0 {
+		todoList = append(todoList, ApplyNodes)
+	}
+	todoList = append(todoList, Guest)
+	todoList = append(todoList, UnMountImage)
+
+	return c.run(todoList)
+}
+
+func (c *Applier) initCluster() error {
+	var todoList []ActionName
+	todoList = append(todoList, PluginDump)
+	todoList = append(todoList, PluginPhaseOriginallyRun)
+	todoList = append(todoList, PullIfNotExist)
+	todoList = append(todoList, MountImage)
+	todoList = append(todoList, Config)
+	todoList = append(todoList, MountRootfs)
+	todoList = append(todoList, PluginPhasePreInitRun)
+	todoList = append(todoList, Init)
+	todoList = append(todoList, PluginPhasePreInstallRun)
+	c.MastersToJoin = c.ClusterDesired.Spec.Masters.IPList[1:]
+	c.NodesToJoin = c.ClusterDesired.Spec.Nodes.IPList
+	todoList = append(todoList, ApplyMasters)
+	todoList = append(todoList, ApplyNodes)
+	todoList = append(todoList, Guest)
+	todoList = append(todoList, UnMountImage)
+	todoList = append(todoList, PluginPhasePostInstallRun)
+	return c.run(todoList)
+}
+
+func (c *Applier) deleteCluster() error {
+	var todoList []ActionName
+	c.MastersToDelete = c.ClusterDesired.Spec.Masters.IPList
+	c.NodesToDelete = c.ClusterDesired.Spec.Nodes.IPList
+	todoList = append(todoList, Reset)
+	todoList = append(todoList, UnMountRootfs)
+	todoList = append(todoList, UnMountImage)
+	todoList = append(todoList, CleanFS)
+	return c.run(todoList)
+}
+
+func (c *Applier) run(todoList []ActionName) error {
+	for _, action := range todoList {
+		logger.Debug("sealer apply process %s", action)
+		err := ActionFuncMap[action](c)
+		if err != nil {
+			return err
+		}
+		if action == Upgrade {
+			c.ClusterCurrent.Spec.Image = c.ClusterDesired.Spec.Image
+		}
```

Comment:
L319-L321 can be moved to Upgrade func (at L80).I realized thing today and I will do it later